### ISSUE TITLE
Actuator configuration questions

### ISSUE THREAD

**Issue by kimbell** [SteeltoeOSS/Steeltoe#504]
I'm trying to get my head around how the actuators are configured. There are a lot of interfaces and several levels of base classes. Is there a particular reason for this level of abstraction? 

Looking at the code, it looks like it's possible to register multiple `IManagementOptions` instances.  Either of type `ActuatorManagementOptions` or `CloudFoundryManagementOptions`. 

```cs
services.TryAddEnumerable(ServiceDescriptor.Singleton<IManagementOptions>(new ActuatorManagementOptions(config)));
services.TryAddSingleton(provider => provider.GetServices<IManagementOptions>().OfType<ActuatorManagementOptions>().First());
```
```cs
services.TryAddEnumerable(ServiceDescriptor.Singleton<IManagementOptions>(new CloudFoundryManagementOptions(config)));
services.TryAddSingleton(provider => provider.GetServices<IManagementOptions>().OfType<CloudFoundryManagementOptions>().First());
```
```cs
var mgmtOptionsCollection = endpoints.ServiceProvider.GetServices<IManagementOptions>();
var builder = conventionBuilder ?? new EndpointCollectionConventionBuilder();

foreach (var mgmtOptions in mgmtOptionsCollection)
```
Why you would need multiple management options is not something I have figured out yet. 

I also noticed that the option classes are not following the `IOptions` pattern from Microsoft, but rather binds things manually. There are also multiple lines of code registering various singletons in the `IServiceCollection`. 

Since the Options classes are defined in EndpointBase rather than EndpointCore, my initial thought was that this might be related to code running on .NET Core and .NET Framework. Since the Option classes have overloaded constructors where one takes in `IConfiguration`, the code gets a dependency on the new Microsoft.Extensions stuff anyway.  

Is .NET Framework still supported?

If multi platform is still required, Microsoft.Extensions.Configuration runs on both platforms; I did this with a SharePoint app several years ago. Configuration would be in json files rather than web.config; something that was acceptable to us.

Is there a particular reason why `IOptions` and `IOptionsMontitor` was not used? With the latter you would get automatic reloading of configuration when a configuration source changes. This could be useful for some of the configuration properties. Clients of Steeltoe could also use `IServiceCollection.PostConfigure()` to make changes. 

Does this code have it's history from a period where depending on the new `IOptions` model was not an option (pun not intended)?

**Comment by hananiel** [comment]
> Is there a particular reason for this level of abstraction?
We are creating some of these abstractions so users have more flexibility in extending functionality. For the most part you do not have to worry about it unless you are. 
>why you would need multiple management options
When running in Cloud Foundry and may be some other customized platform in future, we need the ability to have the endpoints be served via different context paths with different security settings and names etc. So `info` can be registered under multiple contexts `/actuator/info` and `/cloudfoundryapplication/info` for example. Info's `managementoptions` represent the settings for the context path. In addition Cloud Foundry has a custom security middleware that allows it to work on CF. 
> EndpointBase rather than EndpointCore
You are correct that in 2.x we support the .net framework and core and the common functionality is in base. In 3.x it is core only, but we kept the structure, especially since we are supporting both lines. 
> IOptions 
We use IOptions and IOptionsMonitor in some places and some of the code has a history that preceded its availability. 

Hope that helps. Could you please explain what problem are you trying to solve and perhaps we can suggest the best approach? If you want to suggest changes or improvements, PRs are probably the best way to have a structured discussion around specific suggestions.

**Comment by kimbell** [comment]
Since this and #503 are related (at least from my perspective), I'll try to answer them together. 

This all started when I was trying to get the actuator endpoints working. I added the minimum amount of code needed. Some endpoints returned data, others returned nothing. The next step then was to examine the source code to see if I could figure out why things return data. This is when I got lost in all the abstractions and dived down this rabbit hole. 

There are many valid reasons for abstracting things, but when you don't know the history, things can get fuzzy. 

I'd like to take a crack at improving this part.
